sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -332,8 +332,12 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
         execution.Sample(lb, ub, withReplacement, seed, planLater(child)) :: Nil
       case logical.LocalRelation(output, data) =>
         LocalTableScan(output, data) :: Nil
-      case logical.Limit(IntegerLiteral(limit), child) =>
-        execution.Limit(limit, planLater(child)) :: Nil
+      case logical.Limit(IntegerLiteral(limit), child) => {
+        val perPartitionLimit = execution.PartitionLocalLimit(limit, planLater(child))
+        val globalLimit = execution.PartitionLocalLimit(
+          limit, execution.Exchange(SinglePartition, Nil, perPartitionLimit))
```

Comment:
Good idea.  I think we'll have to add something akin to `partial=True` so that we can handle the per-partition and global limits properly.  I'll push another commit shortly to explore this option.